onnxscript/rewriter/pattern.py

```diff
@@ -504,13 +572,44 @@ def commute(self) -> list[ValuePattern]:
         return [self]
 
 
+def _nodes_in_pattern(outputs: Sequence[ValuePattern]) -> list[ValuePattern]:
+    """Computes inputs and nodes from outputs of a pattern."""
+    node_patterns: list[NodePattern] = []
+
+    def visit(value_patterns: Sequence[ValuePattern]) -> None:
+        for value_pattern in value_patterns:
+            if isinstance(value_pattern, NodeOutputPattern):
+                node_pattern = value_pattern.producer()
+                if node_pattern not in node_patterns:
+                    node_patterns.append(node_pattern)
+                    visit(node_pattern.inputs)
+
+    visit(outputs)
+    node_patterns.reverse()
+    return node_patterns
+
+
+def _compute_uses(nodepatterns: Sequence[NodePattern]) -> None:
+    """Compute the uses of values in the pattern."""
+    for nodepattern in nodepatterns:
+        for index, value in enumerate(nodepattern.inputs):
+            if value is not None:
+                value.uses().append((nodepattern, index))
```

Comment:
mutating uses() seems a little strange...

Reply:
I guess I could add an append_use method ... we can construct the uses only finally, after whole pattern-graph has been constructed.

Reply:
Good question ... you are right, we can keep updating it, as and when new nodes are added, but we won't have it when the value is created first (so, I guess that part is common). Right now, uses() is computed finally ... we could change it to do it incrementally when nodes are added. Let me think. This is not required by the single-output-matcher, only by the multi-output matcher.

Reply:
Done